### ISSUE TITLE
[feature request]Support AVX512F intrinstics to vectorize operations

### ISSUE THREAD

**Issue by MlWoo** [pytorch/pytorch#4825]
Pytorch has written SSE, AVX and AVX2 [intrinstics](https://github.com/pytorch/pytorch/tree/master/aten/src/TH/vector) to vectorize operations on CPU.  Now  AVX-512 instruction sets are more and more widely introduced to Intel CPUs.  It is necessary to add AVx-512F intrinstics  to get better performance.

**Comment by apaszke** [comment]
We should be careful with AVX-512 because PyTorch takes a huge performance hit when the clock speed isn't at max (e.g. I've seen many cases where the `powersave` governor reduces perf by 2x even when running on the GPU). So if e.g. someone does a single `sum()` on the CPU and then only GPU ops, AVX-512 will cause heavy downclocking and we might feel this. A careful benchmark suite will be needed.

**Comment by MlWoo** [comment]
Your concern wakes me up. It is very reasonable and must be paid attention to handle with it. The benchmark WIP. BTW， I notice that [simd dispatcher](https://github.com/pytorch/pytorch/blob/master/aten/src/TH/generic/simd/simd.h#L130) could choose executing path according to the environment variables. The AVX512F will also follow the design rules. I think it is a good idea to let users to choose the  appropriate environment variable to enable or disable the AVX512F.

**Comment by apaszke** [comment]
Yes, I think we have env variables for all instruction sets we support, mostly for testing purposes, and AVX-512 should be no different. Thanks!

**Comment by JackKelly** [comment]
Intel's Ice Lake (10th generation) chips are nearly here for laptops, and these include AVX512. Forgive the newbie question, but how do we enable Pytorch to use AVX512?

**Comment by vincentqb** [comment]
> Intel's Ice Lake (10th generation) chips are nearly here for laptops, and these include AVX512. Forgive the newbie question, but how do we enable Pytorch to use AVX512?

@apaszke -- are there other env variables than  [these ones](https://github.com/pytorch/pytorch/blob/master/setup.py#L3) that would be relevant?

**Comment by apaszke** [comment]
I'm not super up to date with the build system, but IIUC ATen still doesn't use AVX512 instructions (and as pointed out above they are not always a clear win).

**Comment by JackKelly** [comment]
Ah, ok, I can appreciate that enabling AVX512 might decrease performance on a system with a fast GPU. But the use-case I'm considering is perhaps an odd one: I'm interested in running PyTorch on an Ice Lake laptop without an Nvidia GPU.  Would enabling AVX512 be a clear win on such a system?

**Comment by soumith** [comment]
@JackKelly not really. AVX512 instructions cannot be sustained on a high clock speed for very long -- the processor downclocks right away if you send too many of them within a short time frame, probably for thermal reasons. This will probably be much more true for laptop processors.

**Comment by uyongw** [comment]
AVX512 max frequency on Skylake server will can be a few 100's MHz lower than AVX2. Ice lake frequency may also be reduced but not that much. [ IA optimization manual](https://software.intel.com/en-us/articles/intel-sdm) section 2.2.3 discusses SKX behavior and section 2.1.1.5 discusses Icelake.  

On Xeon processors, the frequency reduction happens only on the cores that executes AVX-512 instructions. "... each core frequency is determined independently based on the demand of the workload" ([IA optimization manual](https://software.intel.com/en-us/articles/intel-sdm) 2.2.3).

Actually, with MKL/MKLDNN/FBGEMM integrated, the inter-OP is already accelerated by AVX512. Most of time enabling of AVX512 helps improve performance for OPs with large AVX512 code blocks. And the more of AVX512 ratio the bigger benefit we get.  Though, in case of AVX512 is below 20% usually the net benefit is negative.
